### PR TITLE
#603 added search to inbox and messages

### PR DESCRIPTION
**Pull request by komaz**
Fix for issue #603
- Added `Searchable<T> { Iterable<T> search(String term) }` interface to Inbox and Messages
- Initial implementation in all classes implementing `Inbox` and `Messages`
- Created two integration tests for dynamodb
- Added 2 todos to implement the UI

### CODE REVIEW

**Comment by karato** [comment]
@komaz Many thanks for the PR, let me find a reviewer for it

**Comment by karato** [comment]
@ggajos review it

File: netbout-client/src/main/java/com/netbout/client/RtInbox.java [hunk 1]
```diff
@@ -119,4 +119,9 @@ public Bout bout(final long number) {
             }
         };
     }
+
+    @Override
+    public Iterable<Bout> search(final String term) throws IOException {
+        throw new UnsupportedOperationException("#search()");
```
Comment (ggajos) [inline netbout-client/src/main/java/com/netbout/client/RtInbox.java]: @komaz maybe for now we can just return empty iterable, there is no need to react so strongly about missing implementation.

File: netbout-client/src/main/java/com/netbout/client/RtMessages.java [hunk 2]
```diff
@@ -99,4 +99,9 @@ public long unread() {
             }
         };
     }
+
+    @Override
+    public Iterable<Message> search(final String term) throws IOException {
+        throw new UnsupportedOperationException("#search()");
```
Comment (ggajos) [inline netbout-client/src/main/java/com/netbout/client/RtMessages.java]: @komaz same here

File: netbout-web/src/main/java/com/netbout/dynamo/DyInbox.java [hunk 3]
```diff
@@ -67,6 +69,7 @@
 @Loggable(Loggable.DEBUG)
 @ToString(of = "self")
 @EqualsAndHashCode(of = { "counter", "region", "self", "since" })
+@SuppressWarnings({ "PMD.TooManyMethods", "PMD.ExcessiveImports" })
```
Comment (ggajos) [inline netbout-web/src/main/java/com/netbout/dynamo/DyInbox.java]: @komaz maybe we can add ppds to get rid of those supressions

File: netbout-web/src/main/java/com/netbout/dynamo/DyInbox.java [hunk 4]
```diff
@@ -233,6 +236,17 @@ public Bout apply(final Item item) {
         );
     }
 
+    @Override
+    public Iterable<Bout> search(final String term) throws IOException {
+        final List<Bout> result = new ArrayList<>(16);
+        for (final Bout bout: this.iterate()) {
```
Comment (ggajos) [inline netbout-web/src/main/java/com/netbout/dynamo/DyInbox.java]: @komaz add space before double point

File: netbout-web/src/main/java/com/netbout/dynamo/DyMessages.java [hunk 5]
```diff
@@ -73,7 +75,7 @@
 @Loggable(Loggable.DEBUG)
 @ToString(of = "bout")
 @EqualsAndHashCode(of = { "counter", "region", "bout", "self", "start" })
-@SuppressWarnings("PMD.ExcessiveImports")
+@SuppressWarnings({ "PMD.TooManyMethods", "PMD.ExcessiveImports" })
```
Comment (ggajos) [inline netbout-web/src/main/java/com/netbout/dynamo/DyMessages.java]: @komaz maybe we can add ppds to get rid of those supressions

**Comment by ggajos** [comment]
@komaz few comments

File: netbout-web/src/main/java/com/netbout/dynamo/DyInbox.java (hunk 3 above)
Comment (komaz) [inline netbout-web/src/main/java/com/netbout/dynamo/DyInbox.java]: @ggajos what is ppd?

**Comment by komaz** [comment]
@ggajos added impl in `RtInbox` and `RtMessages`, added missing space before colon in for loops

File: netbout-web/src/main/java/com/netbout/dynamo/DyInbox.java (hunk 3 above)
Comment (ggajos) [inline netbout-web/src/main/java/com/netbout/dynamo/DyInbox.java]: @komaz puzzle (todo) in code that will be converted into task, just look for `@todo` in code for example

**Comment by ggajos** [comment]
@komaz http://www.yegor256.com/2009/03/04/pdd.html

**Comment by komaz** [comment]
@ggajos I knew what pdd is, I just was confused as you wrote "ppd", I thought it mean something else. Do you really think that these classes contain too many methods and it worth putting `@todo` here to split it later?

**Comment by ggajos** [comment]
@komaz yes, for example I found some private static methods there, those could be easily extracted.

**Comment by komaz** [comment]
@ggajos ok

**Comment by ggajos** [comment]
@rultor merge

File: netbout-client/src/main/java/com/netbout/client/cached/CdMessages.java [hunk 6]
```diff
@@ -93,4 +93,17 @@ public Message apply(final Message message) {
             }
         );
     }
+
+    @Override
+    public Iterable<Message> search(final String term) throws IOException {
```
Comment (dmzaytsev) [inline netbout-client/src/main/java/com/netbout/client/cached/CdMessages.java]: @komaz why the `@Cacheable` annotation missing here?

**Comment by dmzaytsev** [comment]
@komaz one question above

**Comment by dmzaytsev** [comment]
@komaz ping

**Comment by dmzaytsev** [comment]
@komaz sorry hadn't seen your commit

**Comment by dmzaytsev** [comment]
@rultor merge

**Comment by dmzaytsev** [comment]
@komaz please fix it
> ERROR: "there are 2 duplicate(s) of the same puzzle: netbout-web/src/main/java/com/netbout/dynamo/DyInbox.java:62-65, netbout-web/src/main/java/com/netbout/dynamo/DyMessages.java:68-71, while maximum 1 duplicate is allowed"

**Comment by dmzaytsev** [comment]
@rultor merge

**Comment by karato** [comment]
@elenavolokhova please, review this ticket for compliance with our [QA rules](http://www.teamed.io/qa.html)

**Comment by karato** [comment]
@rultor deploy now pls

**Comment by elenavolokhova** [comment]
@karato Quality is good here.

**Comment by karato** [comment]
> @karato Quality is good here.

@elenavolokhova many thanks

**Comment by karato** [comment]
@ggajos since quality is good, I just added 10 mins to @dmzaytsev (our architect) in transaction 59451521

**Comment by karato** [comment]
@ggajos just added 10 mins to @elenavolokhova (for QA), payment ID is `59456072`. **19 mins** was added to your account, many thanks for your contribution (payment AP-9T1570751H649552H)! The task took 99 hours and 37 mins.. you're getting extra minutes here (c=4). added +19 to your rating, now it is equal to [+1399](http://www.netbout.com/b/35094?open=rating)

**Pull request merged**